akka-http-server/src/main/scala/endpoints/akkahttp/routing/BasicAuthentication.scala

```diff
@@ -0,0 +1,36 @@
+package endpoints.akkahttp.routing
+
+import java.util.Base64
+
+import akka.http.scaladsl.model.headers.{Authorization, BasicHttpCredentials}
+import akka.http.scaladsl.model.{HttpHeader, HttpResponse, StatusCodes}
+import akka.http.scaladsl.server.Directives
+import endpoints.algebra
+import endpoints.algebra.BasicAuthentication.Credentials
+
+trait BasicAuthentication extends algebra.BasicAuthentication with Endpoints {
+
+  /**
+    * Extracts the credentials from the request headers.
+    * In case of absence of credentials rejects request
+    */
+  private[endpoints] lazy val basicAuthentication: RequestHeaders[Credentials] =
+    Directives.headerValue(extractCredentials)
```

Comment:
Ah, OK, I see.

At the beginning the `play.routing` interpreter was working exactly like this but at some point it was useful to introduce a distinction between a request that does not match an endpoint and a request that matches it but is invalid.

So, basically, the result of processing a request should be something like `Option[Either[Invalid, Success]]`, where the `Option` models whether the incoming request matched the endpoint or not, and the `Either` models failures that might have happened when the request was processed.

That’s surely something that we should cleary specify, but for now a request *matches* an endpoint if its method, segments and query string match the endpoint definition. The notion of *matching* defines how we associate an incoming request to an endpoint of the application. Then, after a request has matched an endpoint definition, its processing can *fail* if some expected headers are absent or if the request entity is invalid.